pkg/master/controller/crdregistration/crdregistration_controller.go

```diff
@@ -120,101 +93,31 @@ func (c *crdRegistrationController) Run(threadiness int, stopCh <-chan struct{})
-	err := c.syncHandler(key.(schema.GroupVersion))
-	if err == nil {
-		// if you had no error, tell the queue to stop tracking history for your key.  This will
-		// reset things like failure counts for per-item rate limiting
-		c.queue.Forget(key)
-		return true
-	}
-
-	// there was a failure so be sure to report it.  This method allows for pluggable error handling
-	// which can be used for things like cluster-monitoring
-	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
-	// since we failed, we should requeue the item to work on later.  This method will add a backoff
-	// to avoid hotlooping on particular items (they're probably still not going to work right away)
-	// and overall controller protection (everything I've done is broken, this controller needs to
-	// calm down or it can starve other useful work) cases.
-	c.queue.AddRateLimited(key)
-
-	return true
-}
-
-func (c *crdRegistrationController) enqueueCRD(crd *apiextensions.CustomResourceDefinition) {
-	c.queue.Add(schema.GroupVersion{Group: crd.Spec.Group, Version: crd.Spec.Version})
-}
-
-func (c *crdRegistrationController) handleVersionUpdate(groupVersion schema.GroupVersion) error {
-	found := false
-	apiServiceName := groupVersion.Version + "." + groupVersion.Group
-
-	// check all CRDs.  There shouldn't that many, but if we have problems later we can index them
-	crds, err := c.crdLister.List(labels.Everything())
-	if err != nil {
-		return err
-	}
-	for _, crd := range crds {
-		if crd.Spec.Version == groupVersion.Version && crd.Spec.Group == groupVersion.Group {
-			found = true
-			break
-		}
-	}
-
-	if !found {
-		c.apiServiceRegistration.RemoveAPIServiceToSync(apiServiceName)
-		return nil
-	}
-
+// TODO(mehdy): pass only group/version instead of whole CRD object? This approach is more versions friendly though.
+func (c *crdRegistration) addCRDAPIService(crd *apiextensions.CustomResourceDefinition) {
+	apiServiceName := crd.Spec.Version + "." + crd.Spec.Group
 	c.apiServiceRegistration.AddAPIServiceToSync(&apiregistration.APIService{
 		ObjectMeta: metav1.ObjectMeta{Name: apiServiceName},
 		Spec: apiregistration.APIServiceSpec{
-			Group:                groupVersion.Group,
-			Version:              groupVersion.Version,
+			Group:                crd.Spec.Group,
+			Version:              crd.Spec.Version,
 			GroupPriorityMinimum: 1000, // CRDs should have relatively low priority
 			VersionPriority:      100,  // CRDs should have relatively low priority
 		},
 	})
+}
+
+func (c *crdRegistration) removeCRDAPIService(crd *apiextensions.CustomResourceDefinition) {
+	c.apiServiceRegistration.RemoveAPIServiceToSync(crd.Spec.Version + "." + crd.Spec.Group)
+}
 
-	return nil
+func (c *crdRegistration) updateCRDAPIService(oldCrd, newCrd *apiextensions.CustomResourceDefinition) {
+	// TODO(mehdy): can group/version change during update?
+	c.removeCRDAPIService(oldCrd)
+	c.addCRDAPIService(newCrd)
```

Comment:
Remove/readd isn't right… that will make the make the APIs flutter in discovery and API handling, which we don't want

Reply:
I don't think before versioning update means anything, this should be no-op with single version as you cannot update group or version of a crd object (AFAIK).

Reply:
Update is what gets called on resync and is the chance for the controller to fix up out of sync state